File: server/controllers/issues.py
from flask import Blueprint, jsonify, request
from models.issues import Issue

issue_bp = Blueprint("issues", __name__)

@issue_bp.route('/getissues', methods=["GET"])
def getIssues():
    try:
        res = Issue.get_issues()
        return jsonify(res)
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500

@issue_bp.route('/getissuebyjob', methods=["POST"])
def getIssueByJob():
    try:
        data = request.get_json()
        job_id = data['jobId']
        return Issue.get_issue_by_job(job_id)
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500
    
@issue_bp.route('/createissue', methods=["POST"])
def createIssue(): 
    try:
        data = request.get_json()
        issue = data['issue']
        job = data.get('job', None)
        return Issue.create_issue(issue, job_id=job)
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500
    
@issue_bp.route('/deleteissue', methods=["POST"])
def deleteIssue():
    try:
        data = request.get_json()
        issue_id = data['issueid']
        # print(issueid)
        res = Issue.delete_issue(issue_id)
        return res
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500
    
@issue_bp.route('/editissue', methods=["POST"])
def editIssue():
    try:
        data = request.get_json()
        issue_id = data['issueid']
        issue_new = data['issuenew']
        res = Issue.edit_issue(issue_id, issue_new)
        return res
    except Exception as e:
        print(f"Unexpected error: {e}")
        return jsonify({"error": "Unexpected error occurred"}), 500
    

File: client/src/model/issues.ts
import { api } from './ports'
import { toast } from './toast'

export interface Issue {
    id: number,
    issue: string
}

export function useGetIssues() {
    return {
        async issues() {
            try {
                const response = await api('getissues') // pass rerun job the Job object and desired printer
                if (response.success === true) {
                    return response.issues
                }
            } catch (error) {
                console.error(error)
                toast.error('An error occurred while fetching issues')
            }
        }
    }
}

export function useCreateIssues() {
    return {
        async createIssue(issue: string) {
            try {
                const response = await api('createissue', { issue })
                if (response) {
                    if (response.success == false) {
                        toast.error(response.message)
                    } else if (response.success === true) {
                        toast.success(response.message)
                    } else {
                        console.error('Unexpected response:', response)
                        toast.error('Failed to create issue. Unexpected response')
                    }
                } else {
                    console.error('Response is undefined or null')
                    toast.error('Failed to create issue. Unexpected response')
                }
            } catch (error) {
                console.error(error)
                toast.error('An error occurred while creating the issue')
            }
        }
    }
}

export function useAssignIssue() {
    return {
        async assign(issueid: number, jobid: number) {
            try {
                const response = await api('assignissue', { issueid, jobid })
                if (response) {
                    if (response.success == false) {
                        toast.error(response.message)
                    } else if (response.success === true) {
                        toast.success(response.message)
                    } else {
                        console.error('Unexpected response:', response)
                        toast.error('Failed to assign issue. Unexpected response')
                    }
                } else {
                    console.error('Response is undefined or null')
                    toast.error('Failed to assign issue. Unexpected response')
                }
            } catch (error) {
                console.error(error)
                toast.error('An error occurred while assigning the issue')
            }
        }
    }
}

export function useDeleteIssue() {
    return {
        async deleteIssue(issue: Issue) {
            const issueid = issue.id
            try {
                const response = await api(`deleteissue`, { issueid })
                if (response) {
                    if (response.success == false) {
                        toast.error(response.message)
                    } else if (response.success === true) {
                        toast.success(response.message)
                    } else {
                        console.error('Unexpected response:', response)
                        toast.error('Failed to delete issue. Unexpected response')
                    }
                } else {
                    console.error('Response is undefined or null')
                    toast.error('Failed to delete issue. Unexpected response')
                }
            } catch (error) {
                console.error(error)
                toast.error('An error occurred while deleting the issue')
            }
        }
    }
}

export function useEditIssue() {
    return {
        async editIssue(issueid: number | undefined, issuenew: string) {
            try {
                const response = await api(`editissue`, { issueid, issuenew })
                if (response) {
                    if (response.success == false) {
                        toast.error(response.message)
                    } else if (response.success === true) {
                        toast.success(response.message)
                    } else {
                        console.error('Unexpected response:', response)
                        toast.error('Failed to edit issue. Unexpected response')
                    }
                } else {
                    console.error('Response is undefined or null')
                    toast.error('Failed to edit issue. Unexpected response')
                }
            } catch (error) {
                console.error(error)
                toast.error('An error occurred while editing the issue')
            }
        }
    }
}
